[WIP] Remove Instruct/Chat versions of models & introduce a new ChatTemplate API, fix Anthropic API

Harsha-Nori opened the pull request:
This PR should significantly reduce the number of user-facing classes we have in Guidance, and reduce subtle bugs introduced by using a mis-specified Chat Template (models currently silently default to the ChatML syntax, which many of the latest models don't adhere to). It should also make it easier for users to add new models to guidance, either via PR or in their own codebases.

Before:

```python
from guidance.models.transformers import Transformers, TransformersChat

class Llama(Transformers):
    pass

# Users have to do this for most chat models in guidance
class LlamaChat(TransformersChat, Llama):
    def get_role_start(self, role_name, **kwargs):
        if role_name == "system":
            return self._system_prefex + "<<SYS>>\n"
        elif role_name == "user":
            if str(self).endswith("\n<</SYS>>\n\n"):
                return ""  
            else:
                return "[INST] "
        else:
            return " "

    def get_role_end(self, role_name=None):
        if role_name == "system":
            return "\n<</SYS>>\n\n"
        elif role_name == "user":
            return " [/INST]"
        else:
            return " "

lm = LlamaChat(path_to_llama)
```

After:
```python
from guidance.models import Transformers

lm = Transformers(path_to_llama) # automagically works
```

If you're using a rare model and the auto import doesn't automatically work...

After pt2:
```python
# users can copy paste from https://huggingface.co/meta-llama/Llama-2-7b-chat-hf/blob/main/tokenizer_config.json#L12 
llama2_template = "{% if messages[0]['role'] == 'system' %}{% set loop_messages = messages[1:] %}{% set system_message = messages[0]['content'] %}{% else %}{% set loop_messages = messages %}{% set system_message = false %}{% endif %}{% for message in loop_messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% if loop.index0 == 0 and system_message != false %}{% set content = '<<SYS>>\\n' + system_message + '\\n<</SYS>>\\n\\n' + message['content'] %}{% else %}{% set content = message['content'] %}{% endif %}{% if message['role'] == 'user' %}{{ bos_token + '[INST] ' + content.strip() + ' [/INST]' }}{% elif message['role'] == 'assistant' %}{{ ' '  + content.strip() + ' ' + eos_token }}{% endif %}{% endfor %}"

lm = Transformers(path_to_llama, chat_template=llama2_template)
```

or, in the worst case for maximal robustness and customizability:
```python
from guidance._chat import ChatTemplate, UnsupportedRoleException

class Llama2ChatTemplate(ChatTemplate):
    template_str = llama2_template

    def get_role_start(self, role_name):
        if role_name == "system":
            return "[INST] <<SYS>>\n"
        elif role_name == "user":
            return "<s>[INST]"
        elif role_name == "assistant":
            return " "
        else:
            raise UnsupportedRoleException(role_name, self)
        
    def get_role_end(self, role_name=None):
        if role_name == "system":
            return "\n<</SYS>"
        elif role_name == "user":
            return " [/INST]"
        elif role_name == "assistant":
            return "</s>"
        else:
            raise UnsupportedRoleException(role_name, self)

lm = Transformers(path_to_llama, chat_template=Llama2ChatTemplate)
```

The first big change is the removal of the `Chat` and `Instruct` mixins, and an introduction of a new `guidance._chat.ChatTemplate` class, which handles the same responsibilities as those mixins used to.

Users can construct a subclass of `ChatTemplate` and pass it to models with a new `chat_template` argument (defaulted to None). 

The way this works for local models is to leverage the `chat_template` property in `huggingface transformers` and `llamacpp`'s GGUF files. When a user tries to load a model, guidance now follows the following order of operations:

1. See if the user passed in a `ChatTemplate` -- if so, use that directly.
2. If the user passed string "chat_template", set that as a "template_str". If the user did not pass in anything, set the "template_str" based on metadata from the huggingface.AutoTokenizer or gguf metadata fields.
3. Check the `template_str` against a local cache in guidance which maintains template converters for the most popular models on huggingface/in llama.cpp. We index this cache based on the actual chat_template string, so any model that uses one of these chat templates -- even if it isn't explicitly listed in the documentation -- will automatically load the right guidance class.
4. If we don't have anything in the cache, try to automatically convert the jinja2 template into the new guidance._chat.ChatTemplate syntax. Warn the user if we attempt this. [NOTE: This is not yet implemented and may come in a future PR.]
5. Default to the `ChatML` syntax, with a warning to the user.

Currently this PR updates the following user facing `guidance.Models` classes:

- Transformers (removing TransformersChat and the Llama/Mistral subclasses)
- LlamaCpp (removing LlamaCppChat and Mistral subclasses)
- Anthropic

For now, `Anthropic` should be representative of how grammarless classes will work. I wanted to start with OpenAI, but many other guidance.models classes inherit from OpenAI, so I'm saving that for later. Also while I was at it I upgraded the `Anthropic` class to use their latest SDK, so `guidance.models.Anthropic` should now work with the latest Claude3 models.

# TODO

A decent amount left to do here. In no particular order...

1. Unrelated to this change, but guidance cannot properly handle `llama3` or `phi-3`'s true tokenizers/chat templates. We need to fix that independently.
2. Add better warnings to users when we fall back to using the ChatML template.
3. Extend support through the rest of the guidance.models classes (mostly just remote ones left, starting with OpenAI). 
4. Write out more templates for popular models in the ChatTemplateCache, and also add an alias system so that we can look up models in the cache by common names (e.g. "llama3").
5. Add a deprecation warning to people trying to use very old models on `Anthropic`.
6. Much more testing and documentation. We should, for example, add documentation on how to import/initialize a new ChatTemplate and use it for your own models.
7. Write the auto-converter from huggingface `jinja2` to guidance ChatTemplate. A battery of unit tests here that compare against the original `transformers.apply_chat_template` method would make this more robust. Can be in a future PR as this is complex logic. A start to this was attempted in https://github.com/guidance-ai/guidance/pull/791 by @ilmarinen, and we could eventually pull this in and expand its coverage.
8. Probably get rid of the folders in `guidance.models.llama_cpp` and `guidance.models.transformers` because we don't need to maintain a bunch of subclasses for them anymore.

Would appreciate any and all feedback, particularly on the logical flow and new user facing (simpler) API. @marcotcr @paulbkoch @slundberg @riedgar-ms @hudson-ai

Harsha-Nori commented on guidance/models/_model.py:
```diff
@@ -382,11 +382,9 @@ def next(self, logits):
-                        ) in (
-                            node.keys()
-                        ):  # we update all the children since the parser knows the full mask
+
+                        # we update all the children since the parser knows the full mask
+                        for byte in node.keys():  
```
some of these changes are just undoing particularly egregious dissections that `black` did...e.g. it's wild that this loop iterator was split into 5 lines!

Harsha-Nori replied:
realizing now that it makes this particular PR much harder to parse though, apologies for that!

riedgar-ms replied:
I'm guessing that that was done by the long trailing comment. That seems to be something that `black` is not overly keen on

riedgar-ms replied:
Did you just shift the comment and re-run `black` ?

Harsha-Nori replied:
Yeah I ended up manually moving trailing comments up a line, but it'd be good to see if there's an automatic way to make black handle it.

Harsha-Nori commented on guidance/_chat.py:
```diff
@@ -0,0 +1,210 @@
+# --------------------------------------------------
+# [05/08/24] https://huggingface.co/meta-llama/Llama-2-7b-chat-hf/blob/main/tokenizer_config.json#L12
+llama2_template = "{% if messages[0]['role'] == 'system' %}{% set loop_messages = messages[1:] %}{% set system_message = messages[0]['content'] %}{% else %}{% set loop_messages = messages %}{% set system_message = false %}{% endif %}{% for message in loop_messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/u ...
+class Llama2ChatTemplate(ChatTemplate):
+    # available_roles = ["system", "user", "assistant"]
+    template_str = llama2_template
+
+    def get_role_start(self, role_name):
+        if role_name == "system":
+            return "[INST] <<SYS>>\n"
+        elif role_name == "user":
+            return "<s>[INST]"
+        elif role_name == "assistant":
+            return " "
+        else:
+            raise UnsupportedRoleException(role_name, self)
+        
+    def get_role_end(self, role_name=None):
+        if role_name == "system":
+            return "\n<</SYS>"
+        elif role_name == "user":
+            return " [/INST]"
+        elif role_name == "assistant":
+            return "</s>"
+        else:
+            raise UnsupportedRoleException(role_name, self)
```
this template is likely an oversimplification of the hf template string, I'll need to debug more and extend this.

riedgar-ms replied:
Don't forget to add the tests you develop during your debugging.

Harsha-Nori commented:
Hmm, test failures are related to needing to authenticate on huggingface to load some model tokenizers. Could just disable these in the tests I suppose?
 
 
 > E               Cannot access gated repo for url https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/resolve/main/config.json.
E               Access to model meta-llama/Meta-Llama-3-8B-Instruct is restricted. You must be authenticated to access it.

riedgar-ms commented on guidance/_chat.py:
```diff
@@ -0,0 +1,210 @@
+        self._cache = {}
+
+    def __getitem__(self, key):
+        key_compact = key.replace(" ", "")
```
Minor point: will this collapse multiple consecutive spaces?

Harsha-Nori replied:
nah it's a good catch, I'm actually not sure we need to be doing this. I didn't want minor differences in jinja formats (which I believe are whitespace agnostic for parts of them) to cause different mappings in the cache, but maybe there are places where an extra space is actually a meaningful difference?

riedgar-ms replied:
That was a question I asked before I caught sight of the actual keys you were using....

riedgar-ms commented on guidance/_chat.py:
```diff
@@ -0,0 +1,210 @@
+# @@@@ ChatML @@@@
+# --------------------------------------------------
+# Note that all grammarless models will default to this syntax, since we typically send chat formatted messages.
+chatml_template = "{% for message in messages %}{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
```
Minor point, but shouldn't this the capitalised, if it's a constant?

riedgar-ms commented on guidance/_chat.py:
```diff
@@ -0,0 +1,210 @@
+# @@@@ Llama-2 @@@@
+# --------------------------------------------------
+# [05/08/24] https://huggingface.co/meta-llama/Llama-2-7b-chat-hf/blob/main/tokenizer_config.json#L12
+llama2_template = "{% if messages[0]['role'] == 'system' %}{% set loop_messages = messages[1:] %}{% set system_message = messages[0]['content'] %}{% else %}{% set loop_messages = messages %}{% set system_message = false %}{% endif %}{% for message in loop_messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/u ...
```
Can this be written a little more comprehensibly, but have the newlines/spaces inserted for human readability stripped out on initialisation?

Harsha-Nori replied:
I don't think I understand jinja2 well enough to know if changes I manually make for readability can compromise intrinsics of the template :|. I'd be a bit scared to do that.

riedgar-ms replied:
The idea would be you'd wrap the readable template in something like `textwrap.dedent()`. You could even add a test to verify that it has been converted to what it's supposed to be. `assert string_from_file == "What the string should be"`

Harsha-Nori replied:
I think we can do a round of aesthetics updates after this PR -- we should further extend the ChatTemplate class to have a nice __repr__ for example. I'll leave off these types of changes from this code change.

riedgar-ms replied:
Fair enough given time constraints

riedgar-ms commented on guidance/models/_anthropic.py:
```diff
@@ -52,24 +45,73 @@ def __init__(
+        messages = []
+        found = True
+        system_prompt = None # Not mandatory, but we'll store it if found
+        while found:
```
This block (or one very like it) seems to pop up a lot (I mentioned this when doing the AzureAI Studio examples). Can it be part of the chat template class (or similar)? After all, I believe that it's just 'undoing' the template.

riedgar-ms replied:
Of course it has to be more complicated :-(

riedgar-ms commented on guidance/models/_anthropic.py:
```diff
@@ -52,24 +45,73 @@ def __init__(
         try:
-            generator = self.anthropic.completions.create(
+            # Need to do this because Anthropic API is a bit weird with the system keyword...
+            model_kwargs = dict(
```
Not absolutely necessary, but can you get token stats back from the models as well?

Harsha-Nori replied:
I don't think the Anthropic streaming api supports token usage stats either :( https://docs.anthropic.com/en/api/messages-streaming

Only OpenAI seems to for now.

It might be that each chunk we get sent back is exactly one token? But anthropic's tokenizer is not public so it's also hard to verify this. And there's no way to count the prompt tokens as far as I can tell (without making another API call to a non-streaming version of the API)

riedgar-ms replied:
If they don't support it, they don't support it. If we really wanted approximate metrics, we could grab another tokeniser, but we'd have to warn users that things were approximate. But that's not a discussion for this PR.

riedgar-ms commented on tests/models/test_chat_templates.py:
```diff
@@ -0,0 +1,30 @@
+        "meta-llama/Llama-2-7b-chat-hf", # Llama-2
+        "mistralai/Mistral-7B-Instruct-v0.2", # Mistral-7B-Instruct-v0.2
+    ]
+    for model_id in models_to_test:
```
Please do this via `pytest.parametrize` not a loop inside the test.

Harsha-Nori replied:
done

riedgar-ms reviewed (commented):
Given the changes to `Anthropic`, does that mean we now have access to API keys which we can use in the builds?

Harsha-Nori commented:
> Given the changes to `Anthropic`, does that mean we now have access to API keys which we can use in the builds?

I just made a personal account and paid for $20 of credits out of pocket to test it...not sure it's wise to depend on that for our CI/CD (especially given how frequently they run :| ).

riedgar-ms commented:
> > Given the changes to `Anthropic`, does that mean we now have access to API keys which we can use in the builds?
> 
> I just made a personal account and paid for $20 of credits out of pocket to test it...not sure it's wise to depend on that for our CI/CD (especially given how frequently they run :| ).

I'm quite happy to spend your money.....

riedgar-ms commented:
> Hmm, test failures are related to needing to authenticate on huggingface to load some model tokenizers. I added these new tests in to check the ChatTemplateCache. We could disable these tests, or (better) figure out a way to pass a huggingface auth token here....
> 
> > E               Cannot access gated repo for url https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/resolve/main/config.json.
> > E               Access to model meta-llama/Meta-Llama-3-8B-Instruct is restricted. You must be authenticated to access it.
> 
> @riedgar-ms thoughts here?

You can add the HF auth token to the repo as a secret (like we did for the AzureAI studio secrets), and then access it as an environment variable within the test. I believe I have an `env_or_fail()` method to help with that.

Harsha-Nori commented:
> > Hmm, test failures are related to needing to authenticate on huggingface to load some model tokenizers. I added these new tests in to check the ChatTemplateCache. We could disable these tests, or (better) figure out a way to pass a huggingface auth token here....
> > > E               Cannot access gated repo for url https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/resolve/main/config.json.
> > > E               Access to model meta-llama/Meta-Llama-3-8B-Instruct is restricted. You must be authenticated to access it.
> > 
> > 
> > @riedgar-ms thoughts here?
> 
> You can add the HF auth token to the repo as a secret (like we did for the AzureAI studio secrets), and then access it as an environment variable within the test. I believe I have an `env_or_fail()` method to help with that.

Nice thanks, I changed it to use the env variable and your helper function

riedgar-ms commented on guidance/library/_role.py:
```diff
@@ -25,8 +21,17 @@ def role_opener(lm, role_name, **kwargs):
-
-    lm += lm.get_role_start(role_name, **kwargs)
+    
+    # TODO [HN]: Temporary change while I instrument chat_template in transformers only.
```
Is this still the case (and below)?

riedgar-ms commented on tests/models/test_chat_templates.py:
```diff
@@ -0,0 +1,38 @@
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(model_id, token=hf_token)
+    model_chat_template = tokenizer.chat_template
+    if should_pass:
```
Would separating things out and using pytest's `XFAIL` be better here?

riedgar-ms commented on tests/models/test_openai.py:
```diff
@@ -6,32 +6,32 @@
-            api_key="blah",
-        )
-        assert isinstance(initialized_model, model_class)
+# This is all redundant with the class unification
```
Better deleted than commented out

riedgar-ms commented on tests/models/test_transformers.py:
```diff
@@ -58,27 +58,85 @@ def test_transformer_smoke_select(model_name, model_kwargs):
+    # directly passing in newlines next to special tokens for a tokenizer that does rstrip on those tokens
+    # (like phi-3) will cause a tokenization mismatch issue. 
+    # We're leaving this test in so that we can reliably reproduce and debug this in the future. 
+    with pytest.raises(Exception):
```
Please can you assert something about the exception?

riedgar-ms commented on tests/models/test_transformers.py:
```diff
@@ -58,27 +58,85 @@ def test_transformer_smoke_select(model_name, model_kwargs):
+    assert "5" in lm["five"]
+
+
+@pytest.mark.skip("Don't overload the build machines")
```
I'm getting nervous about the number of tests added which are also skipped. Can you do something like:

https://github.com/guidance-ai/guidance/blob/13270bf9d1d67c3978384cf4df394fb2dfca76f4/tests/models/test_llama_cpp.py#L10-L15

So that these tests will run whenever there's a Phi3 model active in `selected_model` ?

riedgar-ms reviewed (commented):
Have suggested a way of getting some of those Phi3 tests running in the gates

riedgar-ms commented on guidance/models/_openai.py:
```diff
@@ -38,205 +38,50 @@ def __init__(
+                    chunk = ""
+                yield chunk.encode("utf8")
 
+        # Otherwise we are in a chat context
```
Should we have a big 'else' or 'return' here then?

riedgar-ms replied:
Changed

riedgar-ms commented:
@paulbkoch I think that everything is going to pass eventually on this. Per the discussion thread, I'll have to see about converting some more models over to the new paradigm
